api/external.go

```diff
@@ -44,6 +45,11 @@ func (a *API) ExternalProviderRedirect(w http.ResponseWriter, r *http.Request) e
 	query := r.URL.Query()
 	providerType := query.Get("provider")
 	scopes := query.Get("scopes")
+	proxy := query.Get("proxy")
+
+	if proxy != "" {
+		ctx = withExternalProxy(ctx, proxy)
+	}
```

Comment:
Just a couple of thoughts on this:

- I'm not sure the name `proxy` sufficiently explains what the purpose of the param is. A better alternative may be something closer to `redirect_via` or `redirect_through`.
- There's a class of security vulnerabilities for [Unvalidated Redirects and Forwards](https://cheatsheetseries.owasp.org/cheatsheets/Unvalidated_Redirects_and_Forwards_Cheat_Sheet.html) that need to be guarded for here. This value needs to be parsed as a URL and then matched to the [`SiteURL` and the other glob patterns](https://github.com/supabase/gotrue/blob/master/conf/configuration.go#L90-L92) before being accepted for processing.
- Depending on the output of the above, either an error should be returned, or the default mechanism be used. (I personally feel the latter is the right choice.)

Reply:
> There's a class of security vulnerabilities for [Unvalidated Redirects and Forwards](https://cheatsheetseries.owasp.org/cheatsheets/Unvalidated_Redirects_and_Forwards_Cheat_Sheet.html) that need to be guarded for here. This value needs to be parsed as a URL and then matched to the [SiteURL and the other glob patterns](https://github.com/supabase/gotrue/blob/master/conf/configuration.go#L90-L92) before being accepted for processing.

I mentioned this in a response below. But should probably add it here where there's more context.

It's the OAuth provider that redirects to the proxy URL. Supabase hands it to the OAuth provider instead of `https://<project-ref>.supabase.co/auth/v1/callback`. So the proxy URL needs to be added as an authorized URL with the OAuth provider, I'm not sure there's any additional value in Supabase itself performing validations (from a security perspective) - although, I'm more than happy to be told I'm wrong about this!

Reply:
> I'm not sure the name proxy sufficiently explains what the purpose of the param is. A better alternative may be something closer to `redirect_via` or `redirect_through`.

Sorry, forgot to address this. I'm not in any way attached to the name `proxy`. Happy to make this change 👍 The only thing is _technically_ the URL doesn't _need_ to redirect. It could be a reverse proxy to Supabase to save on the redirect - but yeah, your average user really isn't going to want to set that up.

Reply:
> It's the OAuth provider that redirects to the proxy URL. Supabase hands it to the OAuth provider instead of `https://<project-ref>.supabase.co/auth/v1/callback`. So the proxy URL needs to be added as an authorized URL with the OAuth provider, I'm not sure there's any additional value in Supabase itself performing validations (from a security perspective) - although, I'm more than happy to be told I'm wrong about this!

The assumption is that the system is secure, which often opens up the door for security vulnerabilities. It's often the case when building auth software that you need to take extra precautions everywhere. The practice is called _Defense in Depth_ where you layer multiple independent checks for any assumption.

Imagine a misconfigured or vulnerable OAuth provider: a user has forgotten a redirect domain or has a typo in it. Now it's easy to misconfigure the OAuth provider, but it's not easy to change the hosting location of your Supabase project. So the existing solution has 2 checks for the misconfigured OAuth case.

By introducing the proxy you need to add one additional compensating check, as a guard to a misconfigured OAuth provider. Because with the proxy you can now change (and even from a frontend call) the "hosting URL" very easily. You can compensate for this by validating the URL that GoTrue receives. If it is a trusted URL then you can forward the request to the OAuth provider, otherwise you do nothing. Both the OAuth misconfiguration  and the GoTrue misconfiguration must align for an attacker to abuse the system.